src/index/CompressedRelation.cpp

```diff
@@ -1423,11 +1446,61 @@ template <typename T>
 CompressedRelationMetadata CompressedRelationWriter::addCompleteLargeRelation(
     Id col0Id, T&& sortedBlocks) {
   DistinctIdCounter distinctCol1Counter;
-  for (auto& block : sortedBlocks) {
+
+  // Buffer used to ensure the invariant that equal triples (when disregarding
+  // the graph) stay in the same block.
+  std::optional<IdTable> bufferedBlock;
+
+  for (auto& block :
+       sortedBlocks | ql::views::filter(std::not_fn(&IdTable::empty))) {
     ql::ranges::for_each(block.getColumn(1), std::ref(distinctCol1Counter));
-    addBlockForLargeRelation(
-        col0Id, std::make_shared<IdTable>(std::move(block).toDynamic()));
+
+    if (!bufferedBlock.has_value()) {
+      // First non-empty block - initialize buffer.
+      bufferedBlock = std::move(block);
+      continue;
+    }
+
+    const auto& lastRowFromPrevious = bufferedBlock.value().back();
+
+    // Find how many rows from current block have the same first three columns
+    // as the last row in the buffered block
+    const size_t upperBoundEqualTriples =
+        ql::ranges::find_if(
+            block,
+            [&lastRowFromPrevious](const auto& row) {
+              return tieFirstThreeColumns(lastRowFromPrevious) !=
+                     tieFirstThreeColumns(row);
+            }) -
+        block.begin();
+
+    // If we found rows to merge, add them to the buffered block
+    if (upperBoundEqualTriples > 0) {
+      bufferedBlock->insertAtEnd(block, 0, upperBoundEqualTriples);
+
+      // Remove the merged rows from the current block
+      block.erase(block.begin(), block.begin() + upperBoundEqualTriples);
+    }
+
+    // If the `block` is empty after moving the duplicate triples into the
+    // buffer, continue without writing a block, because the next block might
+    // again contain the `lastRowFromPrevious`
+    if (block.empty()) {
+      continue;
+    }
+
+    // At this point we know that the `block` contains at least a single triple
+    // larger than `lastRowFromPrevious`, so we can safely write the
+    // `bufferedBlock`.
+    addBlockForLargeRelation(col0Id, std::move(*bufferedBlock));
+    bufferedBlock = std::move(block);
   }
+
+  // Write the remaining triples from the buffer.
+  if (bufferedBlock.has_value() && !bufferedBlock->empty()) {
```

Comment:
```suggestion
  if (bufferedBlock.has_value()) {
    AD_CORRECTNESS_CHECK(!bufferedBlock->empty());
```

Since empty blocks are filtered, this should always be true I'm 99% sure.

Reply:
I applied this, but I think we currently also don't have a test case for a completely empty relation, so this might still be yellow.